rust/plugin-lib/src/base_cache.rs

```diff
@@ -112,6 +112,14 @@ impl Cache for ChunkCache {
 
     }
 
+    fn get_document<DS: DataSource>(&mut self, source: &DS) -> Result<String, Error> {
```

Comment:
If we're actually going to implement this in the cache itself, it is probably worth doing it a bit more efficiently, e.g.:

```rust
    fn get_document<DS: DataSource>(&mut self, source: &DS) -> Result<String, Error> {
        let mut result = String::new();
        let mut cur_idx = 0;
        while cur_idx < self.buf_size {
            if self.contents.len() == 0 || cur_idx != self.offset {
                let resp = source.get_data(cur_idx, TextUnit::Utf8,
                                           CHUNK_SIZE, self.rev)?;
                self.reset_chunk(resp);
            }
            result.push_str(&self.contents);
            cur_idx = self.offset + self.contents.len();
        }
        Ok(result)
    }
```

The solution here, of doing it one line at a time, would make more sense if you were just wanted to fetch the whole document from within your plugin.

Also ideally this should be returning a `Cow`, since we frequently already have the entire document in the cache and don't need to allocate at all. So maybe you could also add a `//TODO:` mentioning that?